lib/spack/spack/util/package_hash.py

```diff
@@ -74,82 +83,206 @@ def visit_ClassDef(self, node):  # noqa
 
 
 class TagMultiMethods(ast.NodeVisitor):
-    """Tag @when-decorated methods in a spec."""
+    """Tag @when-decorated methods in a package AST."""
     def __init__(self, spec):
         self.spec = spec
+        # map from function name to (implementation, condition_list) tuples
         self.methods = {}
 
-    def visit_FunctionDef(self, node):  # noqa
-        nodes = self.methods.setdefault(node.name, [])
-        if node.decorator_list:
-            dec = node.decorator_list[0]
+    def visit_FunctionDef(self, func):  # noqa
+        conditions = []
+        for dec in func.decorator_list:
             if isinstance(dec, ast.Call) and dec.func.id == 'when':
                 try:
+                    # evaluate spec condition for any when's
                     cond = dec.args[0].s
-                    nodes.append(
-                        (node, self.spec.satisfies(cond, strict=True)))
+
+                    # Boolean literals come through like this
+                    if isinstance(cond, bool):
+                        conditions.append(cond)
+                        continue
+
+                    # otherwise try to make a spec
+                    try:
+                        cond_spec = spack.spec.Spec(cond)
+                    except Exception:
+                        # Spec parsing failed -- we don't know what this is.
+                        conditions.append(None)
+                    else:
+                        # Check statically whether spec satisfies the condition
+                        conditions.append(self.spec.satisfies(cond_spec, strict=True))
+
                 except AttributeError:
                     # In this case the condition for the 'when' decorator is
                     # not a string literal (for example it may be a Python
-                    # variable name). Therefore the function is added
-                    # unconditionally since we don't know whether the
-                    # constraint applies or not.
-                    nodes.append((node, None))
-        else:
-            nodes.append((node, None))
+                    # variable name). We append None because we don't know
+                    # whether the constraint applies or not, and it should be included
+                    # unless some other constraint is False.
+                    conditions.append(None)
+
+        # anything defined without conditions will overwrite prior definitions
+        if not conditions:
+            self.methods[func.name] = []
+
+        # add all discovered conditions on this node to the node list
+        impl_conditions = self.methods.setdefault(func.name, [])
+        impl_conditions.append((func, conditions))
+
+        # don't modify the AST -- return the untouched function node
+        return func
```

Comment:
I'm a little concerned with how this will play with multimethods on inherited packages. Can you make sure to add a test for parsing a package that inherits from a package with multimethod and overrides, and another for adding a condition to the multimethod?

Reply:
So, at the moment there are no inherited multimethods in Spack mainline (either in `build_systems` or in any package that's imported from another), and base packages aren't included in the package hash at all.

There is a question of where to draw the line for the package hash.  I think eventually we should consider making specs even finer-grained and include the spack version or commit in their hashes.  At the same time we might want to include all bases up to but not including `PackageBase`, but then that overlaps with the spack hash in terms of what's represented in a spec.  We could include anything that's *not* in Spack itself, I suppose.

The multimethod stuff is separate because whether we preserve subclass multi methods can depend on whether we're overriding a multimethod from a parent class.  I think that analysis should get factored in, but currently we don't have that information in the AST analysis -- we would need to look up the parent class and parse its AST, as well.

So ultimately I think it's important to handle all this, but I think we should do it in a separate PR.

Reply:
I think it's okay to push this to another PR, but that does mean we will need to have another hash-breaking PR in the future.

Reply:
I think this becomes increasingly less impactful the sooner we make reuse the default.

Reply:
It will still cause a CI backlog at least, even if users are insulated from it